File: scrapers/rk9.py
from .base import BaseScraper
from bs4 import BeautifulSoup
import time
import requests
from typing import Dict, Any, List, Optional
from common.logging import setup_logging

logger = setup_logging()


class RK9Scraper(BaseScraper):
    def __init__(self, db, request_delay: float = 1.0):
        super().__init__(db)
        self.request_delay = request_delay
        self.base_url = "https://rk9.gg"

    def scrape(self, **kwargs) -> Dict[str, Any]:
        tournament_url = kwargs.get('url')
        if not tournament_url:
            logger.error("RK9 tournament URL not provided")
            return {"success": False, "error": "URL required"}

        self._wait_for_rate_limit()

        try:
            logger.info("Scraping RK9 tournament", url=tournament_url)

            tournament_data = self._scrape_tournament_page(tournament_url)
            if not tournament_data:
                logger.error("Failed to scrape tournament page")
                return {"success": False, "error": "Failed to scrape tournament page"}

            tournament_id = tournament_data.get("id") or self._generate_tournament_id(tournament_url)

            if self.tournament_exists(tournament_id):
                logger.info("Tournament already exists, skipping", id=tournament_id)
                return {"success": True, "skipped": True, "tournament_id": tournament_id}

            self._save_tournament(tournament_id, tournament_data)

            roster_url = tournament_url.rstrip('/') + "/roster/"
            self._wait_for_rate_limit()
            self._scrape_roster(roster_url, tournament_id)

            pairings_url = tournament_url.rstrip('/') + "/pairings/"
            self._wait_for_rate_limit()
            self._scrape_pairings(pairings_url, tournament_id)

            logger.info("Successfully scraped RK9 tournament", id=tournament_id)

            return {
                "success": True,
                "tournament_id": tournament_id
            }

        except Exception as e:
            logger.error("Error scraping RK9 tournament", error=str(e), url=tournament_url)
            return {"success": False, "error": str(e)}

    def _wait_for_rate_limit(self):
        time.sleep(self.request_delay)

    def _generate_tournament_id(self, url: str) -> str:
        return f"rk9-{url.rstrip('/').split('/')[-1]}"

    def _scrape_tournament_page(self, url: str) -> Optional[Dict[str, Any]]:
        try:
            response = requests.get(url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')

            name = None
            date = None
            location = None
            format_str = "reg f"

            name_elem = soup.find('h1')
            if name_elem:
                name = name_elem.get_text(strip=True)

            date_elem = soup.find(string=lambda text: 'Date:' in text)
            if date_elem:
                date_text = date_elem.parent.get_text(strip=True).replace('Date:', '').strip()
                date = self._parse_date(date_text)

            location_elem = soup.find(string=lambda text: 'Location:' in text)
            if location_elem:
                location = location_elem.parent.get_text(strip=True).replace('Location:', '').strip()

            if not name:
                return None

            generation, format_name = self.parse_format(format_str)

            return {
                "name": name,
                "date": date,
                "location": location,
                "generation": generation,
                "format": format_name
            }

        except Exception as e:
            logger.error("Error parsing tournament page", error=str(e), url=url)
            return None

    def _scrape_roster(self, url: str, tournament_id: str):
        try:
            response = requests.get(url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')

            player_entries = soup.find_all('div', class_=lambda x: x and 'player' in x.lower())

            for entry in player_entries:
                player_name = None
                country = None
                pokemon_list = []

                name_elem = entry.find(class_=lambda x: x and 'name' in x.lower())
                if name_elem:
                    name_text = name_elem.get_text(strip=True)
                    if '(' in name_text and ')' in name_text:
                        parts = name_text.rsplit('(', 1)
                        player_name = parts[0].strip()
                        country = parts[1].rstrip(')').strip()
                    else:
                        player_name = name_text

                pokemon_elems = entry.find_all(class_=lambda x: x and 'pokemon' in x.lower() or 'poke' in x.lower())
                for poke_elem in pokemon_elems:
                    pokemon_name = poke_elem.get_text(strip=True)
                    if pokemon_name:
                        pokemon_list.append(pokemon_name)

                if player_name:
                    player_id = self.get_or_create_player(player_name, country)
                    team_id = self.get_or_create_team(player_id, tournament_id)

                    for pokemon in pokemon_list:
                        self._save_pokemon_set(team_id, pokemon)

                    logger.debug("Saved player team", player=player_name, pokemon_count=len(pokemon_list))

        except Exception as e:
            logger.error("Error parsing roster", error=str(e), url=url)

    def _scrape_pairings(self, url: str, tournament_id: str):
        try:
            response = requests.get(url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')

            round_elems = soup.find_all('div', class_=lambda x: x and 'round' in x.lower())

            cursor = self.db.conn.cursor()

            for round_elem in round_elems:
                round_text = round_elem.get_text(strip=True)
                round_num = self._extract_round_number(round_text)

                matchups = round_elem.find_all('div', class_=lambda x: x and 'match' in x.lower() or 'pair' in x.lower())

                for matchup in matchups:
                    players = matchup.find_all(class_=lambda x: x and 'player' in x.lower() or 'name' in x.lower())

                    if len(players) >= 2:
                        player1_name = players[0].get_text(strip=True)
                        player2_name = players[1].get_text(strip=True)

                        cursor.execute("""
                            INSERT INTO matches (tournament_id, round_number, table_number)
                            VALUES (?, ?, ?)
                        """, (tournament_id, round_num, None))
                        match_id = cursor.lastrowid

                        player1_id = self.get_or_create_player(player1_name)
                        team1_id = self.get_or_create_team(player1_id, tournament_id)

                        player2_id = self.get_or_create_player(player2_name)
                        team2_id = self.get_or_create_team(player2_id, tournament_id)

                        cursor.execute("""
                            INSERT INTO match_participants (match_id, player_id, team_id, score)
                            VALUES (?, ?, ?, ?)
                        """, (match_id, player1_id, team1_id, 0))

                        cursor.execute("""
                            INSERT INTO match_participants (match_id, player_id, team_id, score)
                            VALUES (?, ?, ?, ?)
                        """, (match_id, player2_id, team2_id, 0))

            self.db.conn.commit()
            logger.debug("Saved pairings", tournament_id=tournament_id)

        except Exception as e:
            logger.error("Error parsing pairings", error=str(e), url=url)

    def _save_tournament(self, tournament_id: str, data: Dict[str, Any]):
        cursor = self.db.conn.cursor()
        cursor.execute("""
            INSERT INTO tournaments (id, name, date, location, generation, format, official)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """, (
            tournament_id,
            data.get("name"),
            data.get("date"),
            data.get("location"),
            data.get("generation"),
            data.get("format"),
            True
        ))
        self.db.conn.commit()

    def _save_pokemon_set(self, team_id: int, pokemon_name: str):
        cursor = self.db.conn.cursor()
        cursor.execute("""
            INSERT INTO pokemon_sets (team_id, species)
            VALUES (?, ?)
        """, (team_id, pokemon_name))
        self.db.conn.commit()

    def _parse_date(self, date_str: str) -> Optional[str]:
        return date_str

    def _extract_round_number(self, round_text: str) -> int:
        import re
        match = re.search(r'\d+', round_text)
        if match:
            return int(match.group())
        return 1

File: cli/src/scrapers/rk9.ts
import { BaseScraper } from './base.js';
import { logger } from '@vgc/common/logging.js';
import { DB } from '../database/db.js';
import axios from 'axios';
import * as cheerio from 'cheerio';

export interface RK9ScraperOptions {
  requestDelay?: number;
}

export class RK9Scraper extends BaseScraper {
  private requestDelay: number;

  constructor(db: DB, options: RK9ScraperOptions = {}) {
    super(db);
    this.requestDelay = options.requestDelay || 1.0;
  }

  async scrape(params: Record<string, unknown>): Promise<Record<string, unknown>> {
    const tournamentUrl = params.url as string | undefined;
    if (!tournamentUrl) {
      logger.error('RK9 tournament URL not provided');
      return { success: false, error: 'URL required' };
    }

    await this.waitForRateLimit();

    try {
      logger.info('Scraping RK9 tournament: url=%s', tournamentUrl);

      const tournamentData = await this.scrapeTournamentPage(tournamentUrl);
      if (!tournamentData) {
        logger.error('Failed to scrape tournament page');
        return { success: false, error: 'Failed to scrape tournament page' };
      }

      const tournamentId = (tournamentData.id && typeof tournamentData.id === 'string') 
        ? tournamentData.id 
        : this.generateTournamentId(tournamentUrl);

      if (this.tournamentExists(tournamentId)) {
        logger.info('Tournament already exists, skipping: id=%s', tournamentId);
        return { success: true, skipped: true, tournament_id: tournamentId };
      }

      this.saveTournament(tournamentId, tournamentData);

      const rosterUrl = tournamentUrl.replace(/\/$/, '') + '/roster/';
      await this.waitForRateLimit();
      await this.scrapeRoster(rosterUrl, tournamentId);

      const pairingsUrl = tournamentUrl.replace(/\/$/, '') + '/pairings/';
      await this.waitForRateLimit();
      await this.scrapePairings(pairingsUrl, tournamentId);

      logger.info('Successfully scraped RK9 tournament: id=%s', tournamentId);

      return {
        success: true,
        tournament_id: tournamentId,
      };
    } catch (error) {
      logger.error('Error scraping RK9 tournament: url=%s error=%s', tournamentUrl, String(error));
      return { success: false, error: String(error) };
    }
  }

  private async waitForRateLimit(): Promise<void> {
    await new Promise(resolve => setTimeout(resolve, this.requestDelay * 1000));
  }

  private generateTournamentId(url: string): string {
    return `rk9-${url.replace(/\/$/, '').split('/').pop()}`;
  }

  private async scrapeTournamentPage(url: string): Promise<Record<string, unknown> | null> {
    try {
      const response = await axios.get(url);
      const $ = cheerio.load(response.data);

      let name: string | null = null;
      let date: string | null = null;
      let location: string | null = null;
      const formatStr = 'reg f';

      const nameElem = $('h1').first();
      if (nameElem.length) {
        name = nameElem.text().trim();
      }

      $('*').each((_, elem) => {
        const text = $(elem).contents().filter((_, n) => n.type === 'text').text();
        if (text && text.includes('Date:')) {
          date = text.replace('Date:', '').trim();
        }
        if (text && text.includes('Location:')) {
          location = text.replace('Location:', '').trim();
        }
      });

      if (!name) {
        return null;
      }

      const { generation, format } = this.parseFormat(formatStr);

      return {
        name,
        date,
        location,
        generation,
        format,
      };
    } catch (error) {
      logger.error({ error: String(error), url }, 'Error parsing tournament page');
      return null;
    }
  }

  private async scrapeRoster(url: string, tournamentId: string): Promise<void> {
    try {
      const response = await axios.get(url);
      const $ = cheerio.load(response.data);

      const playerEntries = $('div[class*="player"]');

      for (let i = 0; i < playerEntries.length; i++) {
        const entry = $(playerEntries[i]);
        let playerName: string | null = null;
        let country: string | undefined;
        const pokemonList: string[] = [];

        const nameElem = entry.find('[class*="name"]').first();
        if (nameElem.length) {
          const nameText = nameElem.text().trim();
          const parenIdx = nameText.lastIndexOf('(');
          if (parenIdx !== -1 && nameText.includes(')')) {
            playerName = nameText.substring(0, parenIdx).trim();
            country = nameText.substring(parenIdx + 1).replace(')', '').trim();
          } else {
            playerName = nameText;
          }
        }

        const pokemonElems = entry.find('[class*="pokemon"], [class*="poke"]');
        pokemonElems.each((_, elem) => {
          const pokemonName = $(elem).text().trim();
          if (pokemonName) {
            pokemonList.push(pokemonName);
          }
        });

        if (playerName) {
          const playerId = this.getOrCreatePlayer(playerName, country);
          const teamId = this.getOrCreateTeam(playerId, tournamentId);

          for (const pokemon of pokemonList) {
            this.savePokemonSet(teamId, pokemon);
          }

          logger.debug({ player: playerName, pokemonCount: pokemonList.length }, 'Saved player team');
        }
      }
    } catch (error) {
      logger.error({ error: String(error), url }, 'Error parsing roster');
    }
  }

  private async scrapePairings(url: string, tournamentId: string): Promise<void> {
    try {
      const response = await axios.get(url);
      const $ = cheerio.load(response.data);

      const roundElems = $('div[class*="round"]');

      for (let i = 0; i < roundElems.length; i++) {
        const roundElem = $(roundElems[i]);
        const roundText = roundElem.text().trim();
        const roundNum = this.extractRoundNumber(roundText);

        const matchups = roundElem.find('div[class*="match"], div[class*="pair"]');

        for (let j = 0; j < matchups.length; j++) {
          const matchup = $(matchups[j]);
          const players = matchup.find('[class*="player"], [class*="name"]');

          if (players.length >= 2) {
            const player1Name = $(players[0]).text().trim();
            const player2Name = $(players[1]).text().trim();

            const matchResult = this.db.prepare(`
              INSERT INTO matches (tournament_id, round_number, table_number)
              VALUES (?, ?, ?)
            `).run(tournamentId, roundNum, null);
            const matchId = matchResult.lastInsertRowid;

            const player1Id = this.getOrCreatePlayer(player1Name);
            const team1Id = this.getOrCreateTeam(player1Id, tournamentId);

            const player2Id = this.getOrCreatePlayer(player2Name);
            const team2Id = this.getOrCreateTeam(player2Id, tournamentId);

            this.db.prepare(`
              INSERT INTO match_participants (match_id, player_id, team_id, score)
              VALUES (?, ?, ?, ?)
            `).run(matchId, player1Id, team1Id, 0);

            this.db.prepare(`
              INSERT INTO match_participants (match_id, player_id, team_id, score)
              VALUES (?, ?, ?, ?)
            `).run(matchId, player2Id, team2Id, 0);
          }
        }
      }

      logger.debug({ tournamentId }, 'Saved pairings');
    } catch (error) {
      logger.error({ error: String(error), url }, 'Error parsing pairings');
    }
  }

  private saveTournament(tournamentId: string, data: Record<string, unknown>): void {
    this.db.prepare(`
      INSERT INTO tournaments (id, name, date, location, generation, format, official)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(
      tournamentId,
      data.name,
      data.date,
      data.location,
      data.generation,
      data.format,
      1
    );
  }

  private savePokemonSet(teamId: number, pokemonName: string): void {
    this.db.prepare(`
      INSERT INTO pokemon_sets (team_id, species)
      VALUES (?, ?)
    `).run(teamId, pokemonName);
  }

  private extractRoundNumber(roundText: string): number {
    const match = roundText.match(/\d+/);
    if (match) {
      return parseInt(match[0], 10);
    }
    return 1;
  }
}
